File: src/demo-operations.mjs
import { boxVertices } from './object-identity.mjs';
import { ensureMaterial } from './material-operations.mjs';
import { applyTransform, normalizeTransform, normalizeVector, positiveNumber } from './operation-utils.mjs';
import { addWarning, boundingBoxForVertices } from './snapshot.mjs';

export function addDemoRoom(model, operation = {}) {
  const width = positiveNumber(operation.width, 4500, 'room.width');
  const depth = positiveNumber(operation.depth, 3000, 'room.depth');
  const height = positiveNumber(operation.height, 2400, 'room.height');
  const wallThickness = positiveNumber(operation.wall_thickness ?? operation.wallThickness, 120, 'room.wall_thickness');
  const floorThickness = positiveNumber(operation.floor_thickness ?? operation.floorThickness, 100, 'room.floor_thickness');
  const name = operation.name || 'Demo_Room';

  ensureMaterial(model, 'Floor_Oak', '#a87945');
  ensureMaterial(model, 'Wall_Paint', '#efe7dc');
  ensureMaterial(model, 'Door_Wood', '#7a4a2b');
  ensureMaterial(model, 'Window_Glass', '#8ecae6');
  ensureMaterial(model, 'Table_Wood', '#9b6b43');
  ensureMaterial(model, 'Chair_Fabric', '#315c8a');

  addDemoBox(model, {
    name: `${name}_Floor`,
    origin: [0, 0, 0],
    size: [width, depth, floorThickness],
    material: 'Floor_Oak'
  });

  const doorWidth = 900;
  const doorHeight = 2100;
  const doorX = (width - doorWidth) / 2;
  const wallZ = floorThickness;

  addDemoBox(model, {
    name: `${name}_Wall_South_Left`,
    origin: [0, -wallThickness, wallZ],
    size: [doorX, wallThickness, height],
    material: 'Wall_Paint'
  });
  addDemoBox(model, {
    name: `${name}_Wall_South_Right`,
    origin: [doorX + doorWidth, -wallThickness, wallZ],
    size: [width - doorX - doorWidth, wallThickness, height],
    material: 'Wall_Paint'
  });
  addDemoBox(model, {
    name: `${name}_Wall_South_Header`,
    origin: [doorX, -wallThickness, wallZ + doorHeight],
    size: [doorWidth, wallThickness, height - doorHeight],
    material: 'Wall_Paint'
  });
  addDemoBox(model, {
    name: `${name}_Door_Panel`,
    origin: [doorX + 25, -wallThickness - 25, wallZ],
    size: [doorWidth - 50, 25, doorHeight],
    material: 'Door_Wood'
  });

  addDemoBox(model, {
    name: `${name}_Wall_North`,
    origin: [0, depth, wallZ],
    size: [width, wallThickness, height],
    material: 'Wall_Paint'
  });
  addDemoBox(model, {
    name: `${name}_Wall_West`,
    origin: [-wallThickness, 0, wallZ],
    size: [wallThickness, depth, height],
    material: 'Wall_Paint'
  });
  addDemoBox(model, {
    name: `${name}_Wall_East`,
    origin: [width, 0, wallZ],
    size: [wallThickness, depth, height],
    material: 'Wall_Paint'
  });

  addDemoBox(model, {
    name: `${name}_Window_North_Glass`,
    origin: [width * 0.58, depth + wallThickness + 6, wallZ + 1050],
    size: [1050, 12, 750],
    material: 'Window_Glass'
  });

  const tableX = width / 2 - 600;
  const tableY = depth / 2 - 400;
  addDemoBox(model, { name: `${name}_Table_Top`, origin: [tableX, tableY, 750], size: [1200, 800, 75], material: 'Table_Wood' });
  for (const [index, leg] of [[0, 0], [1100, 0], [0, 700], [1100, 700]].entries()) {
    addDemoBox(model, { name: `${name}_Table_Leg_${index + 1}`, origin: [tableX + leg[0], tableY + leg[1], 100], size: [100, 100, 650], material: 'Table_Wood' });
  }

  addDemoBox(model, { name: `${name}_Chair_Seat`, origin: [tableX + 300, tableY - 550, 450], size: [600, 500, 75], material: 'Chair_Fabric' });
  addDemoBox(model, { name: `${name}_Chair_Back`, origin: [tableX + 300, tableY - 600, 525], size: [600, 75, 700], material: 'Chair_Fabric' });
  addDemoBox(model, { name: `${name}_Chair_Leg_1`, origin: [tableX + 350, tableY - 500, 100], size: [75, 75, 350], material: 'Chair_Fabric' });
  addDemoBox(model, { name: `${name}_Chair_Leg_2`, origin: [tableX + 775, tableY - 500, 100], size: [75, 75, 350], material: 'Chair_Fabric' });
  addDemoBox(model, { name: `${name}_Chair_Leg_3`, origin: [tableX + 350, tableY - 150, 100], size: [75, 75, 350], material: 'Chair_Fabric' });
  addDemoBox(model, { name: `${name}_Chair_Leg_4`, origin: [tableX + 775, tableY - 150, 100], size: [75, 75, 350], material: 'Chair_Fabric' });

  addWarning(model, 'info.limitation', 'info', 'Door opening is represented by segmented wall boxes in the MVP DSL.', 'demo_room');
  addWarning(model, 'info.limitation', 'info', 'Window is represented as glass marker geometry; true boolean wall cuts are left for the SketchUp plugin refinement.', 'demo_room');
}

function addDemoBox(model, operation) {
  const { name, origin, size, material } = operation;
  const normalizedOrigin = normalizeVector(origin, [0, 0, 0], `${name}.origin`);
  const normalizedSize = normalizeVector(size, [1, 1, 1], `${name}.size`);
  if (normalizedSize.some((value) => value <= 0)) throw new Error(`${name}.size values must be positive`);
  const vertices = demoBoxVertices(normalizedOrigin, normalizedSize);
  const transformedVertices = applyTransform(vertices, operation, name);
  model.groups.push({
    id: name,
    name,
    kind: 'box',
    faces: 6,
    edges: 12,
    material: ensureMaterial(model, material),
    transform: normalizeTransform(operation, name),
    bounding_box: boundingBoxForVertices(transformedVertices),
    _vertices: transformedVertices
  });
}

function demoBoxVertices(origin, size) {
  return boxVertices(origin, size);
}

File: sketchup_plugin/alma_sketchup_mcp/demo_operations.rb
# frozen_string_literal: true

module AlmaSketchupMCP
  extend self

  def add_demo_room(entities, operation)
    width = positive_number(operation['width'], 4500, 'room.width')
    depth = positive_number(operation['depth'], 3000, 'room.depth')
    height = positive_number(operation['height'], 2400, 'room.height')
    wall_thickness = positive_number(operation['wall_thickness'] || operation['wallThickness'], 120, 'room.wall_thickness')
    floor_thickness = positive_number(operation['floor_thickness'] || operation['floorThickness'], 100, 'room.floor_thickness')
    name = operation['name'] || 'Demo_Room'

    ensure_material('Floor_Oak', '#a87945')
    ensure_material('Wall_Paint', '#efe7dc')
    ensure_material('Door_Wood', '#7a4a2b')
    ensure_material('Window_Glass', '#8ecae6')
    ensure_material('Table_Wood', '#9b6b43')
    ensure_material('Chair_Fabric', '#315c8a')

    add_box(entities, 'name' => "#{name}_Floor", 'origin' => [0, 0, 0], 'size' => [width, depth, floor_thickness], 'material' => 'Floor_Oak')

    door_width = 900.0
    door_height = 2100.0
    door_x = (width - door_width) / 2.0
    wall_z = floor_thickness

    add_box(entities, 'name' => "#{name}_Wall_South_Left", 'origin' => [0, -wall_thickness, wall_z], 'size' => [door_x, wall_thickness, height], 'material' => 'Wall_Paint')
    add_box(entities, 'name' => "#{name}_Wall_South_Right", 'origin' => [door_x + door_width, -wall_thickness, wall_z], 'size' => [width - door_x - door_width, wall_thickness, height], 'material' => 'Wall_Paint')
    add_box(entities, 'name' => "#{name}_Wall_South_Header", 'origin' => [door_x, -wall_thickness, wall_z + door_height], 'size' => [door_width, wall_thickness, height - door_height], 'material' => 'Wall_Paint')
    add_box(entities, 'name' => "#{name}_Door_Panel", 'origin' => [door_x + 25, -wall_thickness - 25, wall_z], 'size' => [door_width - 50, 25, door_height], 'material' => 'Door_Wood')
    add_box(entities, 'name' => "#{name}_Wall_North", 'origin' => [0, depth, wall_z], 'size' => [width, wall_thickness, height], 'material' => 'Wall_Paint')
    add_box(entities, 'name' => "#{name}_Wall_West", 'origin' => [-wall_thickness, 0, wall_z], 'size' => [wall_thickness, depth, height], 'material' => 'Wall_Paint')
    add_box(entities, 'name' => "#{name}_Wall_East", 'origin' => [width, 0, wall_z], 'size' => [wall_thickness, depth, height], 'material' => 'Wall_Paint')
    add_box(entities, 'name' => "#{name}_Window_North_Glass", 'origin' => [width * 0.58, depth + wall_thickness + 6, wall_z + 1050], 'size' => [1050, 12, 750], 'material' => 'Window_Glass')

    table_x = width / 2.0 - 600
    table_y = depth / 2.0 - 400
    add_box(entities, 'name' => "#{name}_Table_Top", 'origin' => [table_x, table_y, 750], 'size' => [1200, 800, 75], 'material' => 'Table_Wood')
    [[0, 0], [1100, 0], [0, 700], [1100, 700]].each_with_index do |leg, index|
      add_box(entities, 'name' => "#{name}_Table_Leg_#{index + 1}", 'origin' => [table_x + leg[0], table_y + leg[1], 100], 'size' => [100, 100, 650], 'material' => 'Table_Wood')
    end
    add_box(entities, 'name' => "#{name}_Chair_Seat", 'origin' => [table_x + 300, table_y - 550, 450], 'size' => [600, 500, 75], 'material' => 'Chair_Fabric')
    add_box(entities, 'name' => "#{name}_Chair_Back", 'origin' => [table_x + 300, table_y - 600, 525], 'size' => [600, 75, 700], 'material' => 'Chair_Fabric')
    add_box(entities, 'name' => "#{name}_Chair_Leg_1", 'origin' => [table_x + 350, table_y - 500, 100], 'size' => [75, 75, 350], 'material' => 'Chair_Fabric')
    add_box(entities, 'name' => "#{name}_Chair_Leg_2", 'origin' => [table_x + 775, table_y - 500, 100], 'size' => [75, 75, 350], 'material' => 'Chair_Fabric')
    add_box(entities, 'name' => "#{name}_Chair_Leg_3", 'origin' => [table_x + 350, table_y - 150, 100], 'size' => [75, 75, 350], 'material' => 'Chair_Fabric')
    add_box(entities, 'name' => "#{name}_Chair_Leg_4", 'origin' => [table_x + 775, table_y - 150, 100], 'size' => [75, 75, 350], 'material' => 'Chair_Fabric')

    add_warning('info.limitation', 'info', 'Door opening is represented by segmented wall boxes in the MVP DSL.', 'demo_room')
    add_warning('info.limitation', 'info', 'Window is represented as glass marker geometry; true boolean wall cuts are left for refinement.', 'demo_room')
  end
end
